New resource: azurerm_dev_test_global_shutdown_schedule

sean-nixon opened the pull request:
This resource enables the "Auto-shutdown" functionality of regular ARM-based VMs. It's basically a hack of the Dev Test Labs schedules resource known as a "Global Schedule" in the API docs. The API looks like it would support multiple types of schedules and timings, but in reality based on testing only "ComputeVMShutdownTask" schedules are supported and only with a daily recurrence. The API also seems to require a specific naming standard of the resource of "shutdown-computevm-{{VM Name}}". Anything else results in a 400 Bad Request.

Because of these restrictions, for ease of use, I opted to create a resource specifically for shutdown schedules that essentially maps to the "Auto-shutdown" blade in the Portal. If in the future, if additional types of schedules like Auto-starts are supported, my thought is new resources could be created or a more general resource could be created to support multiple types.

I'm very open to feedback on resource naming and overall approach.

**References:**

Feature Announcement
https://azure.microsoft.com/en-us/blog/announcing-auto-shutdown-for-vms-using-azure-resource-manager/

API Docs
https://docs.microsoft.com/en-us/rest/api/dtl/globalschedules

Example Usage Forum Post
https://social.msdn.microsoft.com/Forums/en-US/25a02403-dba9-4bcb-bdcc-1f4afcba5b65/powershell-script-to-autoshutdown-azure-virtual-machine?forum=WAVirtualMachinesforWindows

sean-nixon commented:
@tombuildsstuff or @katbyte I know y'all have been busy prepping for the v2.0 release, but would love to get feedback on this PR when you get the chance. I'm working on setting up auto-shutdown on VMs for a client right now and we'd love to have a native Terraform resource to do so.

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["schedules"]
```
Could we write a parse ID function like seen in #5779 ?

sean-nixon replied:
Added a function for this as suggested.

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"time": {
```
Would this make more sense as a top level field that is: `daily_recurrence_time `

sean-nixon replied:
I left it like this to match the existing `dev_test_schedule` resource as well as the structure of the API but I think I agree with making it a top level attribute. I left the `flatten` and `expand` functions in place and updated them to handle a simple string instead. Would appreciate feedback on that approach.

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+
+func testAccAzureRMDevTestLabGlobalShutdownSchedule_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
```
we'll need a provider block for all tests now

sean-nixon replied:
Done.

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+func testAccAzureRMDevTestLabGlobalShutdownSchedule_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
and could we use
```suggestion
  name     = "acctestRG-dtl-%d"
```

sean-nixon replied:
Done

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+resource "azurerm_subnet" "test" {
+  name                 = "acctsub-%d"
```
all resource names should contain acctest and caps when possible
```suggestion
  name                 = "acctest-SN-%d"
```

sean-nixon replied:
Makes sense. I updated all resources (where possible) to use that naming format.

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+)
+
+func TestAccAzureRMDevTestLabGlobalShutdownSchedule_autoShutdownBasic(t *testing.T) {
```
Could we have 3 tests: basic, complete, update
- basic is just the required properties
- complete is everything possible set
- update is basic -> complete -> basic

sean-nixon replied:
Added the additional tests as you suggested.

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  tags = {
+    environment = "Production"
```
Make sure caps work for keys
```suggestion
    Environment = "Production"
```

sean-nixon replied:
Done.

katbyte commented on website/docs/r/dev_test_global_shutdown_schedule.html.markdown:
```diff
@@ -0,0 +1,140 @@
+  }
+}
+
+resource "azurerm_virtual_machine" "sample" {
```
Could we change this to the new linux/windows VM resource?

sean-nixon replied:
Done.

katbyte commented on azurerm/internal/services/devtestlabs/tests/resource_arm_dev_test_global_shutdown_schedule_test.go:
```diff
@@ -0,0 +1,243 @@
+  }
+}
+
+resource "azurerm_virtual_machine" "test" {
```
Could we use the new windows/linux vm resource here?

sean-nixon replied:
Done.

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+			},
+
+			"status": {
+				Type:     schema.TypeString,
```
Updated to use boolean

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+				},
+			},
+
+			"time_zone_id": {
```
I agree. Updated accordingly

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"status": {
```
Done.

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+	defer cancel()
+
+	targetResourceID := d.Get("target_resource_id").(string)
+	rid, err := azure.ParseAzureResourceID(targetResourceID)
```
Added custom parse and validate functions. Would love feedback on them.

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+		Schema: map[string]*schema.Schema{
+			"location": azure.SchemaLocation(),
+
+			"target_resource_id": {
```
It's a normal virtual machine ID, not a dev test VM. I updated it to be `virtual_machine_id` like with extensions which is a little clearer. Do you agree with that approach?

sean-nixon commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,324 @@
+			TargetResourceID: &targetResourceID,
+			TaskType:         &taskType,
+		},
+		Tags: tags.Expand(t),
```
Done.

sean-nixon commented:
@katbyte Thanks for reviewing this! Appreciate all your hard work. I just got to making the updates you suggested. I also made some additional minor tweaks to the schema to address some issues that came up during my testing:
- Made `notification_settings.enabled` required to make things simpler and not have scenarios with an empty `notification_settings` block
- Removed validation from `notification_settings.webhook_url` as I was unable to find an equivalent for `validate.NoEmptyStrings`
- Removed `Computed: true` from `notification_settings.webhook_url` to ensure that Terraform properly detects a change if the URL is set on the resource in Azure but not in the configuration

I think this should be ready for another round of review when you get the chance.

sean-nixon commented:
@katbyte Is there anything else I can address on this PR?

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,303 @@
+		return err
+	}
+
+	vmName := vm.Name
+	resGroup := vm.ResourceGroup
+
```
I don't think we need to do this? we can use vm.Name where we use that var (or idealy id.Name for consistancy)

sean-nixon replied:
Removed

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,303 @@
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	vmID := d.Get("virtual_machine_id").(string)
+	vm, err := parse.GlobalScheduleVirtualMachineID(vmID)
```
this should be
```suggestion
	id, err := parse.GlobalScheduleVirtualMachineID(d.Get("virtual_machine_id").(string))
```

sean-nixon replied:
Updated to `id` instead of `vm` as suggested. However, I left the `vmID` variable intact because I also require it later when constructing the request to create the schedule. I thought that preferable to duplicating `d.Get("virtual_machine_id").(string)` when constructing the request.

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,303 @@
+		Schema: map[string]*schema.Schema{
+			"location": azure.SchemaLocation(),
+
+			"virtual_machine_id": {
```
it seems like this resource acts on a vm? not globally as the name implies?

should this resource instead be called `azurerm_dev_test_vm_global_shutdown_schedule` to highlight it works on a vm?

sean-nixon replied:
The global part comes from the underlying API. I think it's named that because it actually operates outside the context of DevTestLabs, but the API is still under the Microsoft.DevTestLabs provider. Thus it's a "global" schedule in the sense that it's not local to DevTestLabs.

I'm really not a fan of the name as is honestly. In my opinion, ideally the resource name wouldn't even have `dev_test` in it because it's not really applicable to DevTestLabs, it just uses the same resource provider. In an ideal world, I would name it `azurerm_virtual_machine_shutdown_schedule` so that it would be more easily discoverable and intuitive to find. I know that if I was looking through the documentation and saw a resource `azurerm_dev_test_vm_global_shutdown_schedule` I would not think that it would apply to normal Azure compute VMs, but it does.

I don't know the project's appetite for breaking the namespacing of resources like that. If that's too much, I'm fine with your proposed resource name. I might just propose switching the order to `global_vm_shutdown_schedule` to make "global" parse as a modifier for "vm" to better indicate it's not referencing normal DevTestLabs VMs.

I will work on the other changes you've requested to clean up the code a bit. Once that's done I can update the name however you decide.

katbyte replied:
@sean-nixon - entirely agree, the API and SDK naming can be pretty awful. We break it all the time for what will make the most sense, and typically try to follow the portal over the API. i think `azurerm_dev_test_global_vm_shutdown_schedule` is a good compromise, as it is part of the dev test labs service.

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,303 @@
+	if err != nil {
+		return err
+	}
+	resGroup := schedule.ResourceGroup
+	name := schedule.Name
```
again we don't need to assign these tolocal alues

sean-nixon replied:
Removed

katbyte commented on azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go:
```diff
@@ -0,0 +1,303 @@
+	if err != nil {
+		return err
+	}
+	resGroup := schedule.ResourceGroup
+	name := schedule.Name
```
and here

sean-nixon replied:
Removed

sean-nixon commented:
Hi @katbyte, thanks for the great suggestions. I think I've cleaned up everything from your latest review. I think the last remaining question is the resource name. Let me know your thoughts on my inline comment above and I'll refactor the resource name accordingly.

katbyte commented:
@sean-nixon - not sure if you saw my comment reply - but i'm thinking `azurerm_dev_test_global_vm_shutdown_schedule` is a great compromise/final name to run with!

sean-nixon commented:
Hi @katbyte, thank you for reminding me to come back to this. I did see your comment earlier but hadn't gotten the chance to come back and refactor everything. I just pushed up my latest changes to use the new name. I also updated the documentation for the new resource to try to contrast it more clearly with the regular `dev_test_schedule` resource and to call out the resource ID format for imports. Let me know if you have any objections to those changes.

I also validated that the integration tests are still working. 

```
$ make acctests SERVICE='devtestlabs' TESTARGS='-run=TestAccAzureRMDevTestLabGlobalVMShutdownSchedule'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
TF_ACC=1 go test -v ./azurerm/internal/services/devtestlabs/tests/ -run=TestAccAzureRMDevTestLabGlobalVMShutdownSchedule -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownBasic
=== PAUSE TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownBasic
=== RUN   TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownComplete
=== PAUSE TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownComplete
=== RUN   TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownUpdate
=== PAUSE TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownUpdate
=== CONT  TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownBasic
=== CONT  TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownUpdate
=== CONT  TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownComplete
--- PASS: TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownBasic (431.16s)
--- PASS: TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownComplete (555.11s)
--- PASS: TestAccAzureRMDevTestLabGlobalVMShutdownSchedule_autoShutdownUpdate (623.81s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/devtestlabs/tests   624.863s
```

Let me know if there's anything else you'd like for me to address. Would love to see this in the next release 😊

sean-nixon commented:
@stefan-rapp I would probably recommend creating an issue for this. The reason I didn't include email is because it would require an SDK update and I didn't want to try to introduce that together with my change as I'm not familiar with the process. Once the SDK version is updated, adding email support should be simple.

stefan-rapp commented:
@sean-nixon : https://github.com/terraform-providers/terraform-provider-azurerm/issues/7221#issuecomment-639375647
